### PR TITLE
Finish the CouchDB test matrix

### PR DESCRIPTION
**Pull request by jeqq**

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -9,16 +9,25 @@ mysql:
...
 
+services:
+  - couchdb
+
 env:
-  - PATH=$PATH:/home/travis/.composer/vendor/bin
+  matrix:
+    - SCRIPT='modules/relaxed/tests/bin/drupal.sh'
+    - SCRIPT='modules/relaxed/tests/bin/drupal_to_couchdb.sh'
+    - SCRIPT='modules/relaxed/tests/bin/couchdb_to_drupal.sh'
```
Comment (dickolsson) [inline .travis.yml]: I prefer to use `-` in filenames instead of `_`, so let's do that instead :)

File: .travis.yml [hunk 2]
```diff
@@ -9,16 +9,25 @@ mysql:
...
 
-before_install:
-  # Update Composer so that the Drush installation doesn't fail.
-  - composer selfupdate
+matrix:
+  allow_failures:
+    - env: SCRIPT='modules/relaxed/tests/bin/drupal_to_couchdb.sh'
+    - env: SCRIPT='modules/relaxed/tests/bin/couchdb_to_drupal.sh'
```
Comment (dickolsson) [inline .travis.yml]: Filenames need updating here as well.

File: relaxed.module [hunk 3]
```diff
@@ -50,3 +52,12 @@ function relaxed_entity_delete(EntityInterface $entity) {
...
 }
+
+/**
+ *  Implements hook_entity_create_access().
+ */
+function relaxed_entity_create_access(AccountInterface $account, array $context, $entity_bundle) {
+  if ($entity_bundle == 'replication_log') {
+    return AccessResult::allowedIfHasPermission($account, 'restful put relaxed:local:doc');
+  }
+}
```
Comment (dickolsson) [inline relaxed.module]: Let's convert this to a proper access handler instead. That would be more flexible. So we need to create a handler and add it to the `handler` array in [the annotation for ReplicationLog](https://github.com/dickolsson/drupal-relaxed/blob/8.x-1.x/src/Entity/ReplicationLog.php#L17).

I suggest that we create a very simple access handler, something very similar to `\Drupal\entity_test\EntityTestAccessControlHandler`. I really like the idea of reusing the `restful put relaxed:local:doc` permission. So we should keep that in this new access handler.

File: src/Plugin/rest/resource/EnsureFullCommitResource.php [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+ * Contains \Drupal\relaxed\Plugin\rest\resource\EnsureFullCommitResource.
+ */
+
+namespace Drupal\relaxed\Plugin\rest\resource;
+
+use Drupal\rest\ResourceResponse;
+use Symfony\Component\HttpKernel\Exception\BadRequestHttpException;
+
+/**
+ * @RestResource(
```
Comment (dickolsson) [inline src/Plugin/rest/resource/EnsureFullCommitResource.php:18]: Let's add a comment line about this resource, something like:

``` php
/**
 * This resource does not do anything useful since Drupal does not (yet) have a
 * concept of transactions across multiple requests. This resource only exists so
 * that the API is compliant with the replication protocol.
 *
 * @RestResource(
 * ...
 * )
 */
```

File: relaxed.module (hunk 3 above)
Comment (dickolsson) [inline relaxed.module]: Just a side note for now; I added a new issue to the roadmap related to this: https://www.drupal.org/node/2385701

File: src/Normalizer/ContentEntityNormalizer.php [hunk 5]
```diff
@@ -105,8 +105,14 @@ public function denormalize($data, $class, $format = NULL, array $context = arra
     $entity_id = NULL;
 
     // Get the entity type and the entity id from $data['_id'] string.
-    if (!empty($data['_id'])) {
-      list($entity_type_from_data, $entity_id_from_data) = explode(':', $data['_id']);
+    if (!empty($data['_id']) && strpos($data['_id'], '.') !== false) {
```
Comment (dickolsson) [inline src/Normalizer/ContentEntityNormalizer.php]: This should use uppercase `FALSE` as per Drupal standards.

File: src/Normalizer/ContentEntityNormalizer.php [hunk 6]
```diff
@@ -105,8 +105,14 @@ public function denormalize($data, $class, $format = NULL, array $context = arra
...
+      list($entity_type_from_data, $entity_id_from_data) = explode('.', $data['_id']);
+    }
+    elseif (!empty($data['_id']) && strpos($data['_id'], '/') !== false) {
```
Comment (dickolsson) [inline src/Normalizer/ContentEntityNormalizer.php]: This should use uppercase `FALSE` as per Drupal standards.

File: tests/bin/drupal-to-couchdb.sh [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+
+# Create a target database and do the replication.
+curl -X PUT localhost:5984/target
+curl -X GET http://admin:admin@localhost/relaxed/default
+nohup curl -X POST -H "Accept: application/json" -H "Content-Type: application/json" -d '{"source": "http://admin:admin@localhost/relaxed/default", "target": "http://localhost:5984/target"}' http://localhost:5984/_replicate &
+sleep 300
+
+# Output information useful for debugging.
+sudo cat /var/log/couchdb/couch.log
+sudo cat /var/log/apache2/forensic.log
```
Comment (dickolsson) [inline tests/bin/drupal-to-couchdb.sh]: I don't think we need these `cat` commands anymore. They're mostly useful for debugging. But now when we have solved all problems we don't need them to clutter the Travis log anymore.

File: tests/bin/couchdb-to-drupal.sh [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+
+curl -X GET http://admin:admin@localhost/relaxed/default/1111b3b1-6d76-4813-9e12-18d4e91e1111
+curl -X GET http://admin:admin@localhost/relaxed/default/2222b3b1-6d76-4813-9e12-18d4e91e2222
+curl -X GET http://admin:admin@localhost/relaxed/default/3333b3b1-6d76-4813-9e12-18d4e91e3333
+
+# Output information from couch.log.
+sudo cat /var/log/couchdb/couch.log
+
+# Output information from forensic.log.
+sudo cat /var/log/apache2/forensic.log
```
Comment (dickolsson) [inline tests/bin/couchdb-to-drupal.sh]: Same here, I think we can drop these `cat` commands now.

File: src/Normalizer/ContentEntityNormalizer.php [hunk 9]
```diff
@@ -179,8 +185,16 @@ public function denormalize($data, $class, $format = NULL, array $context = arra
...
     }
 
+    // Add _rev_info field info to the $data array.
+    if (isset($data['_rev']) && isset($data['_revisions']['start']) && isset($data['_revisions']['ids'])) {
+      $parts = explode('-', $data['_rev']);
+      if ($parts[0] == $data['_revisions']['start'] && in_array($parts[1], $data['_revisions']['ids'])) {
+        $data['_revs_info'][0]['rev'] = $data['_rev'];
+      }
+    }
+
```
Comment (dickolsson) [inline src/Normalizer/ContentEntityNormalizer.php:202]: We should only add `_revs_info` to the data if the `revs_info` or the `open_revs` query parameter is available. See the spec: http://docs.couchdb.org/en/1.6.1/api/document/common.html#get--db-docid

I already implemented some logic for this in the field normalizer for the `_revs_info` field, see `\Drupal\relaxed\Normalizer\RevisionInfoItemListNormalizer`. We should probably move the above code into that specific field normalizer to isolate it better.

We can use the `$context` variable for this, [like we do for _revisions](https://github.com/dickolsson/drupal-relaxed/blob/8.x-1.x/src/Normalizer/ContentEntityNormalizer.php#L74). This is possible because [we pass all query parameters into the context in the controller](https://github.com/dickolsson/drupal-relaxed/blob/8.x-1.x/src/Controller/ResourceController.php#L146).

File: src/Normalizer/ContentEntityNormalizer.php [hunk 10]
```diff
@@ -75,7 +75,7 @@ public function normalize($entity, $format = NULL, array $context = array()) {
       $parts = explode('-', $entity->_revs_info->rev);
       $data['_revisions'] = array(
         'ids' => array(),
-        'start' => $parts[0],
+        'start' => (int) $parts[0],
```
Comment (dickolsson) [inline src/Normalizer/ContentEntityNormalizer.php:81]: Is this something that was needed for the replication to work? If so, let's add a test small test for this so that we make sure that this keeps working as expected.

File: src/Plugin/Field/FieldType/ReplicationHistoryItem.php [hunk 11]
```diff
@@ -145,7 +145,7 @@ public static function schema(FieldStorageDefinitionInterface $field_definition)
         ),
         'start_time' => array(
           'type' => 'varchar',
-          'length' => 20,
+          'length' => 50,
```
Comment (dickolsson) [inline src/Plugin/Field/FieldType/ReplicationHistoryItem.php:148]: Let's add more asserts in `\Drupal\relaxed\Tests\ReplicationLogTest` so that we test the data types, lengths etc that the replication protocol expects.

File: src/Normalizer/ContentEntityNormalizer.php [hunk 12]
```diff
@@ -192,8 +206,16 @@ public function denormalize($data, $class, $format = NULL, array $context = arra
...
+      if ($entity) {
+        foreach ($data as $name => $value) {
+          $entity->{$name} = $value;
+        }
+      }
+      elseif (isset($data['id'])) {
+        unset($data['id']);
+        $entity_id = NULL;
+        /** @var \Drupal\Core\Entity\ContentEntityInterface $entity */
+        $entity = $storage->save($data);
```
Comment (dickolsson) [inline src/Normalizer/ContentEntityNormalizer.php]: Why was this added? The normalizer should never be responsible for saving any entities. We need good separation of our code. Can we perform the save operation inside a resource plugin instead?

File: tests/bin/drupal-to-couchdb.sh [hunk 13]
```diff
@@ -0,0 +1,19 @@
...
+curl -X GET http://localhost:5984/target/_all_docs
```
Comment (dickolsson) [inline tests/bin/drupal-to-couchdb.sh]: We need to pipe this last `curl` command to a `grep` command (or something similar) so that we can assert that the right revisions got replicated and exit with `0` or `1` accordingly. Otherwise this test matrix will always pass with green results which is not very useful :)

File: tests/bin/couchdb-to-drupal.sh [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+# Check if default workspace can be accessed.
+curl -X GET http://admin:admin@localhost/relaxed/default
+
+# Run the replication.
+nohup curl -X POST -H "Accept: application/json" -H "Content-Type: application/json" -d '{"source": "http://localhost:5984/source", "target": "http://admin:admin@localhost/relaxed/default"}' http://localhost:5984/_replicate &
+sleep 300
+
+curl -X GET http://admin:admin@localhost/relaxed/default/1111b3b1-6d76-4813-9e12-18d4e91e1111
+curl -X GET http://admin:admin@localhost/relaxed/default/2222b3b1-6d76-4813-9e12-18d4e91e2222
+curl -X GET http://admin:admin@localhost/relaxed/default/3333b3b1-6d76-4813-9e12-18d4e91e3333
```
Comment (dickolsson) [inline tests/bin/couchdb-to-drupal.sh]: We should change these `curl` commands to just be one `curl` command to `http://admin:admin@localhost/relaxed/default/_all_docs` (like we do in `drupal-to-couchdb.sh`). We should then pipe the result to `grep` to assert that the right revisions got replicated and exit with the appropriate status.

File: tests/bin/couchdb-to-drupal.sh [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+
+# Run the replication.
+nohup curl -X POST -H "Accept: application/json" -H "Content-Type: application/json" -d '{"source": "http://localhost:5984/source", "target": "http://admin:admin@localhost/relaxed/default"}' http://localhost:5984/_replicate &
+sleep 300
+
+curl -X GET http://admin:admin@localhost/relaxed/default/_all_docs | tee /tmp/all_docs.txt
+
+# Analyze the output to ascertain the right revisions got replicated.
+ALL_DOCS=$(egrep -c "(2-53af89597b78f87defb3fd9c6845072a)|(3-a751de6677a6196258905cb60851ce0f)|(4-d3b3807b37e52224a8833feab50d8af6)" /tmp/all_docs.txt)
+if [ $ALL_DOCS -eq 3 ]; then exit 0; else exit 1; fi
```
Comment (dickolsson) [inline tests/bin/couchdb-to-drupal.sh]: I like what you did in `drupal-to-couchdb.sh` with just `grep` on `total_rows`. Let's do that here as well. That's simple enough.

File: tests/bin/drupal-to-couchdb.sh [hunk 16]
```diff
@@ -0,0 +1,20 @@
...
+# Create a target database and do the replication.
+curl -X PUT localhost:5984/target
+curl -X GET http://admin:admin@localhost/relaxed/default
+nohup curl -X POST -H "Accept: application/json" -H "Content-Type: application/json" -d '{"source": "http://admin:admin@localhost/relaxed/default", "target": "http://localhost:5984/target"}' http://localhost:5984/_replicate &
+sleep 300
+
+curl -X GET http://localhost:5984/target/_all_docs | tee /tmp/all_docs.txt
+
+# Analyze the output to ascertain the right revisions got replicated.
+ALL_DOCS=$(egrep -c "(\"total_rows\"\:10)|(2-)|(1-)" /tmp/all_docs.txt)
```
Comment (dickolsson) [inline tests/bin/drupal-to-couchdb.sh]: It will be enough to just `grep` for `\"total_rows\"\:10`. I don't think we really need to grep on the revision prefix.

File: src/Normalizer/ContentEntityNormalizer.php (hunk 9 above)
Comment (jeqq) [inline src/Normalizer/ContentEntityNormalizer.php:202]: @dickolsson We need to do this on entity denormalization, and I think it is not possible to do it in `\Drupal\relaxed\Normalizer\RevisionInfoItemListNormalizer`. I can add a check if the `revs_info` or the `open_revs` parameter is available in `\Drupal\relaxed\Normalizer\ContentEntityNormalizer`. What do you think?

Comment (dickolsson) [inline src/Normalizer/ContentEntityNormalizer.php:202]: @jeqq That's right, denormalization for fields is not possible in separate classes. Ok. Yes, please add the check for `revs_info` and `open_revs` inside `\Drupal\relaxed\Normalizer\ContentEntityNormalizer`.

File: tests/bin/couchdb-to-drupal.sh [hunk 17]
```diff
@@ -0,0 +1,45 @@
...
+     -d '{"_id":"entity_test.2222b3b1-6d76-4813-9e12-18d4e91e2222","_rev":"2-a2a2bb67431dd59bf85bdd6d6ff82222","id":[{"value":"2"}],"uuid":[{"value":"2222b3b1-6d76-4813-9e12-18d4e91e2222"}],"langcode":[{"value":"en"}],"name":[{"value":"Entity name 2"}],"type":[{"value":"entity_test"}],"user_id":[{"target_id":"1"}],"revision_id":[{"value":"2"}],"field_test_text":[{"value":null,"format":null}],"workspace":[{"target_id":"target"}],"_revisions":{"start":2,"ids":["35aa41544020dcf1bdcfa664e3760926","1b69807df77d61db1b9efd3742965730"]}}' \
+     localhost:5984/source
+curl -X POST \
+     -H "Content-Type: application/json" \
+     -d '{"_id":"entity_test.3333b3b1-6d76-4813-9e12-18d4e91e3333","_rev":"3-a2a2bb67431dd59bf85bdd6d6ff83333","id":[{"value":"3"}],"uuid":[{"value":"3333b3b1-6d76-4813-9e12-18d4e91e3333"}],"langcode":[{"value":"en"}],"name":[{"value":"Entity name 3"}],"type":[{"value":"entity_test"}],"user_id":[{"target_id":"1"}],"revision_id":[{"value":"3"}],"field_test_text":[{"value":null,"format":null}],"workspace":[{"target_id":"target"}],"_revisions":{"start":3,"ids":["a098b75ab1947368ab7c9a632c0ea02a","35aa41544020dcf1bdcfa674e3760926"]}}' \
+     localhost:5984/source
```
Comment (dickolsson) [inline tests/bin/couchdb-to-drupal.sh]: Let's move the test documents into a separate file (`./tests/fixtures/documents.txt`) with one document per line and make a loop like this instead:

``` bash
while read document
  curl -X POST \
       -H "Content-Type: application/json" \
       -d $document
       localhost:5984/source
done < $TRAVIS_BUILD_DIR/tests/fixtures/documents.txt
```

Then we can do this at the very end:

``` bash
COUNT=$(wc -l $TRAVIS_BUILD_DIR/tests/fixtures/documents.txt)
test 1 -eq $(egrep -c "(\"total_rows\"\:$COUNT)" /tmp/all_docs.txt)
```

I don't think we need the complex `if then exit` logic in the end. I think this will be cleaner. Then it's also easier to add new documents to test with.

File: tests/bin/drupal-to-couchdb.sh [hunk 18]
```diff
@@ -0,0 +1,20 @@
...
+curl -X PUT localhost:5984/target
+curl -X GET http://admin:admin@localhost/relaxed/default
+nohup curl -X POST -H "Accept: application/json" -H "Content-Type: application/json" -d '{"source": "http://admin:admin@localhost/relaxed/default", "target": "http://localhost:5984/target"}' http://localhost:5984/_replicate &
+sleep 300
+
+curl -X GET http://localhost:5984/target/_all_docs | tee /tmp/all_docs.txt
+
+# Analyze the output to ascertain the right revisions got replicated.
+ALL_DOCS=$(egrep -c "(\"total_rows\"\:10)" /tmp/all_docs.txt)
+if [ $ALL_DOCS -eq 1 ]; then exit 0; else exit 1; fi
```
Comment (dickolsson) [inline tests/bin/drupal-to-couchdb.sh]: Let's use the same `test` command logic here as I suggested in the previous comment.

File: tests/modules/relaxed_test/relaxed_test.install [hunk 19]
```diff
@@ -0,0 +1,33 @@
...
+      $entity = entity_create($entity_type);
+      $entity->save();
+      $entity->name = 'Entity name ' . $entity->id();
+      if ($entities_number % 2 == 0) {
+        $entity->delete();
+      }
+      $entity->save();
+      $entities_number--;
+    }
+  }
```
Comment (dickolsson) [inline tests/modules/relaxed_test/relaxed_test.install]: Let's remove the `foreach` loop here and create all documents in `drupal-to-couchdb.sh` with the `while read document` loop and `$TRAVIS_BUILD_DIR/tests/fixtures/documents.txt` as I suggested above.

File: src/Plugin/rest/resource/AllDocsResource.php [hunk 20]
```diff
@@ -0,0 +1,50 @@
...
+   *
+   * @return \Drupal\rest\ResourceResponse
+   */
+  public function get($workspace) {
+    $db = $workspace->id();
+    $docs = array();
+    $multiversion_manager = \Drupal::service('multiversion.manager');
+    $entity_types = \Drupal::service('entity.manager')->getDefinitions();
+    foreach ($entity_types as $entity_type) {
+      if ($multiversion_manager->isSupportedEntityType($entity_type)) {
```
Comment (dickolsson) [inline src/Plugin/rest/resource/AllDocsResource.php]: This endpoint should only return documents that are NOT local. So we also need to check that the entity type is NOT local in this if statement.

File: .travis.yml [hunk 21]
```diff
@@ -9,16 +9,25 @@ mysql:
...
+    - SCRIPT='modules/relaxed/tests/bin/drupal-to-couchdb.sh'
+    - SCRIPT='modules/relaxed/tests/bin/couchdb-to-drupal.sh'
 
-before_install:
-  # Update Composer so that the Drush installation doesn't fail.
-  - composer selfupdate
+matrix:
+  allow_failures:
+    - env: SCRIPT='modules/relaxed/tests/bin/drupal-to-couchdb.sh'
+    - env: SCRIPT='modules/relaxed/tests/bin/couchdb-to-drupal.sh'
```
Comment (dickolsson) [inline .travis.yml]: We should remove `allow_failures` since all tests should be green from now on.

File: src/Normalizer/AllDocsNormalizer.php [hunk 22]
```diff
@@ -0,0 +1,24 @@
...
+    $uuid = $data->uuid();
+    return array(
+      'id' => $uuid,
+      'key' => $uuid,
+      'value' => array(
+        'rev' => $data->_revs_info->rev,
+      ),
+    );
+  }
+}
```
Comment (dickolsson) [inline src/Normalizer/AllDocsNormalizer.php:34]: We need a test class that only tests this normalizer.

File: src/Normalizer/BulkDocsNormalizer.php [hunk 23]
```diff
@@ -21,6 +21,9 @@ public function normalize($data, $format = NULL, array $context = array()) {
         $result['docs'][] = parent::normalize($field, $format, $context);
       }
     }
+    elseif (isset($context['resource_id']) && $context['resource_id'] == 'relaxed:all_docs') {
+      return \Drupal::service('relaxed.normalizer.all_docs')->normalize($data, $format, $context);
+    }
```
Comment (dickolsson) [inline src/Normalizer/BulkDocsNormalizer.php]: This feels very wrong. The all docs normalizer should not be mixed with the bulk docs normalizer. To do it properly they should be in separate classes.

File: src/Normalizer/AllDocsNormalizer.php [hunk 24]
```diff
@@ -0,0 +1,24 @@
...
+  protected $supportedInterfaceOrClass = array('Drupal\Core\Entity\ContentEntityInterface');
+
+  /**
+   * {@inheritdoc}
+   */
+  public function normalize($data, $format = NULL, array $context = array()) {
+    $uuid = $data->uuid();
+    return array(
+      'id' => $uuid,
+      'key' => $uuid,
```
Comment (dickolsson) [inline src/Normalizer/AllDocsNormalizer.php]: Both `id` and `key` needs to be in the same form as all other endpoints, i.e. `[ENTITY TYPE].[UUID]`.

File: src/Tests/Normalizer/ContentEntityNormalizerTest.php [hunk 25]
```diff
@@ -39,10 +39,10 @@ public function testNormalize() {
 
     $expected = array(
       'id' => array(
-        array('value' => 1),
+        array('value' => '1'),
       ),
       'revision_id' => array(
-        array('value' => 1),
+        array('value' => '1'),
```
Comment (dickolsson) [inline src/Tests/Normalizer/ContentEntityNormalizerTest.php]: What's causing this? I would rather fix so that the source returns correctly cases values instead.

File: tests/bin/drupal-to-couchdb.sh [hunk 26]
```diff
@@ -0,0 +1,20 @@
...
+set -ev
+
+# Enable dependencies.
+mv $TRAVIS_BUILD_DIR/../drupal/core/modules/system/tests/modules/entity_test $TRAVIS_BUILD_DIR/../drupal/modules/entity_test
+mv $TRAVIS_BUILD_DIR/../drupal/modules/relaxed/tests/modules/relaxed_test $TRAVIS_BUILD_DIR/../drupal/modules/relaxed_test
+drush en --yes entity_test, relaxed_test || true
+
+# Create a target database and do the replication.
+curl -X PUT localhost:5984/target
+curl -X GET http://admin:admin@localhost/relaxed/default
```
Comment (dickolsson) [inline tests/bin/drupal-to-couchdb.sh]: We don't need this `curl -X GET` command anymore. I think this was only for debugging.

File: src/Normalizer/BulkDocsNormalizer.php (hunk 23 above)
Comment (jeqq) [inline src/Normalizer/BulkDocsNormalizer.php]: @dickolsson It's already in a separate class, we have three normalizers where `$supportedInterfaceOrClass = array('Drupal\Core\Entity\ContentEntityInterface')`: `ContentEntityNormalizer`, `BulkDocsNormalizer` and `AllDocsNormalizer`. Please recommend a better method to handle which class will be used for entity normalization.
